### PR TITLE
[LoRA] feat: support loading regular Flux LoRAs into Flux Control, and Fill

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?
Fixes https://github.com/huggingface/diffusers/issues/10180, https://github.com/huggingface/diffusers/issues/10227, #10184 

In short, this PR enables few-steps inference for Flux Control, Fill, Redux, etc. 

<details>
<summary>Fill + Turbo LoRA</summary>

```py
from diffusers import FluxFillPipeline
from diffusers.utils import load_image
import torch

pipe = FluxFillPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-Fill-dev", torch_dtype=torch.bfloat16
).to("cuda")

adapter_id = "alimama-creative/FLUX.1-Turbo-Alpha"
pipe.load_lora_weights(adapter_id)

image = load_image("https://huggingface.co/datasets/diffusers/diffusers-images-docs/resolve/main/cup.png")
mask = load_image("https://huggingface.co/datasets/diffusers/diffusers-images-docs/resolve/main/cup_mask.png")

image = pipe(
    prompt="a white paper cup",
    image=image,
    mask_image=mask,
    height=1632,
    width=1232,
    guidance_scale=30,
    num_inference_steps=8,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
image.save("flux-fill-dev.png")
```

</details>

<details>
<summary>Flux Control LoRA + Turbo LoRA (different from the previous one)</summary>

```py
from diffusers import FluxControlPipeline
from image_gen_aux import DepthPreprocessor
from diffusers.utils import load_image
from huggingface_hub import hf_hub_download
import torch

control_pipe = FluxControlPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16)
control_pipe.load_lora_weights("black-forest-labs/FLUX.1-Depth-dev-lora", adapter_name="depth")
control_pipe.load_lora_weights(
    hf_hub_download("ByteDance/Hyper-SD", "Hyper-FLUX.1-dev-8steps-lora.safetensors"), adapter_name="hyper-sd"
)
control_pipe.set_adapters(["depth", "hyper-sd"], adapter_weights=[0.85, 0.125])
control_pipe.enable_model_cpu_offload()

prompt = "A robot made of exotic candies and chocolates of different kinds. The background is filled with confetti and celebratory gifts."
control_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/robot.png")

processor = DepthPreprocessor.from_pretrained("LiheYoung/depth-anything-large-hf")
control_image = processor(control_image)[0].convert("RGB")

image = control_pipe(
    prompt=prompt,
    control_image=control_image,
    height=1024,
    width=1024,
    num_inference_steps=8,
    guidance_scale=10.0,
    generator=torch.Generator().manual_seed(42),
).images[0]
image.save("output.png")
```

</details>

## Todods

- [ ] Integration tests
- [x] Docs

### CODE REVIEW

**Comment by sayakpaul** [comment]
Requesting for a review from @BenjaminBossan for initial stuff. Then will request reviews from others.

File: src/diffusers/loaders/lora_pipeline.py [hunk 1]
```diff
@@ -2329,56 +2331,90 @@ def _maybe_expand_transformer_param_shape_or_error_(
...
 
-                # TODO: consider initializing this under meta device for optims.
-                expanded_module = torch.nn.Linear(
-                    in_features, out_features, bias=bias, device=module_weight.device, dtype=module_weight.dtype
-                )
-                # Only weights are expanded and biases are not.
-                new_weight = torch.zeros_like(
-                    expanded_module.weight.data, device=module_weight.device, dtype=module_weight.dtype
+    @classmethod
+    def _maybe_expand_lora_state_dict(cls, transformer, lora_state_dict):
```
Comment (BenjaminBossan) [inline src/diffusers/loaders/lora_pipeline.py:2387]: Just for my understanding, only LoRA is expanded because only the input shape is larger, but output shape stays the same?

File: src/diffusers/loaders/lora_pipeline.py [hunk 2]
```diff
@@ -2329,56 +2331,90 @@ def _maybe_expand_transformer_param_shape_or_error_(
...
-                    attribute_name = _MODULE_NAME_TO_ATTRIBUTE_MAP_FLUX[current_module_name]
-                    new_value = int(expanded_module.weight.data.shape[1])
-                    old_value = getattr(transformer.config, attribute_name)
-                    setattr(transformer.config, attribute_name, new_value)
-                    logger.info(f"Set the {attribute_name} attribute of the model to {new_value} from {old_value}.")
-
-        return has_param_with_shape_update
+        if expanded_module_names:
+            logger.info(
+                f"Found some LoRA modules for which the weights were zero-padded: {expanded_module_names}. Please open an issue if you think this was unexpected - https://github.com/huggingface/diffusers/issues/new."
```
Comment (BenjaminBossan) [inline src/diffusers/loaders/lora_pipeline.py:2415]: I think the message could be a bit misleading, since it's diffusers that did the padding, but it reads like diffusers just discovered this. How about a more active phrasing:

"The following LoRA modules were zero padded to [reason]: {expanded_module_names}.".

Also, could be changed to debug log IMHO, unless there is good reason to be believe it could fire inadvertently.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for extending the functionality of loading LoRA adapters when shapes need to be expanded. The PR LGTM, I only have a nit.

One question that came up (maybe it was already discussed and I just missed it or forgot): Right now, this type of expansion is permanent, right? I.e. even after unloading the LoRA that made the expansion necessary in the first place, the expansion is not undone. Probably that would be quite hard to add and not worth the effort, I'm just curious.

File: src/diffusers/loaders/lora_pipeline.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/loaders/lora_pipeline.py:2387]: Yeah. Only the inputs.

File: src/diffusers/loaders/lora_pipeline.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/loaders/lora_pipeline.py:2428]: Our users prefer explicit logging I believe. So, I logged it as an info. But no strong opinions.

Comment (sayakpaul) [reply src/diffusers/loaders/lora_pipeline.py]: Done fixing the error message.

**Comment by sayakpaul** [comment]
@BenjaminBossan 

> Right now, this type of expansion is permanent, right? I.e. even after unloading the LoRA that made the expansion necessary in the first place, the expansion is not undone. Probably that would be quite hard to add and not worth the effort, I'm just curious.

The LoRA state dict expansion is permanent. But model-level state dict expansion can be undone is being added in https://github.com/huggingface/diffusers/pull/10206

**Comment by BenjaminBossan** [comment]
> The LoRA state dict expansion is permanent. But model-level state dict expansion can be undone is being added in #10206

Yes, something similar for LoRA would be nice, but it's not as important, as the overhead for LoRA should be relatively small.

File: src/diffusers/loaders/lora_pipeline.py [hunk 3]
```diff
@@ -2329,56 +2331,102 @@ def _maybe_expand_transformer_param_shape_or_error_(
...
+                if out_features > module_out_features or in_features > module_in_features:
+                    has_param_with_shape_update = True
+                    parent_module_name, _, current_module_name = name.rpartition(".")
+                    parent_module = transformer.get_submodule(parent_module_name)
+
+                    with torch.device("meta"):
+                        expanded_module = torch.nn.Linear(
+                            in_features, out_features, bias=bias, dtype=module_weight.dtype
+                        )
+                    # Only weights are expanded and biases are not.
```
Comment (a-r-r-o-w) [inline src/diffusers/loaders/lora_pipeline.py:2360]:
```suggestion
                    # Only weights are expanded and biases are not. This is because only the input dimensions
                    # are changed while the output dimensions remain the same. The shape of the weight tensor
                    # is (out_features, in_features), while the shape of bias tensor is (out_features,), which
                    # explains the reason why only weights are expanded.
```

File: tests/lora/test_lora_layers_flux.py [hunk 4]
```diff
@@ -478,21 +463,16 @@ def test_lora_expanding_shape_with_normal_lora_raises_error(self):
             "transformer.x_embedder.lora_B.weight": normal_lora_B.weight,
         }
 
-        # The first lora expanded the input features of x_embedder. Here, we are trying to load a lora with the correct
```
Comment (a-r-r-o-w) [inline tests/lora/test_lora_layers_flux.py:481]: So this error is now removed because expanding lora state dicts is now supported

File: tests/lora/test_lora_layers_flux.py [hunk 5]
```diff
@@ -535,17 +515,107 @@ def test_lora_expanding_shape_with_normal_lora_raises_error(self):
...
         }
 
-        # We should check for input shapes being incompatible here. But because above mentioned issue is
-        # not a supported use case, and because of the PEFT renaming, we will currently have a shape
-        # mismatch error.
+        # We should check for input shapes being incompatible here.
         self.assertRaisesRegex(
             RuntimeError,
-            "size mismatch for x_embedder.lora_A.adapter-2.weight",
+            "x_embedder.lora_A.weight",
```
Comment (a-r-r-o-w) [inline tests/lora/test_lora_layers_flux.py:521]: Why is that part of the comment being removed, and why is `size mismatch` part being removed?

Just to confirm, we still don't support:
- Load N different "normal" loras
- Try to load "shape expander" lora
- Will the N different normal loras be expanded?

File: tests/lora/test_lora_layers_flux.py [hunk 6]
```diff
@@ -535,17 +515,107 @@ def test_lora_expanding_shape_with_normal_lora_raises_error(self):
...
+        self.assertFalse(np.allclose(lora_output, lora_output_2, atol=1e-3, rtol=1e-3))
+        self.assertFalse(np.allclose(lora_output, lora_output_3, atol=1e-3, rtol=1e-3))
+        self.assertFalse(np.allclose(lora_output_2, lora_output_3, atol=1e-3, rtol=1e-3))
+
+        pipe.fuse_lora(lora_scale=1.0, adapter_names=["adapter-1", "adapter-2"])
+        lora_output_4 = pipe(**inputs, generator=torch.manual_seed(0))[0]
+        self.assertTrue(np.allclose(lora_output_3, lora_output_4, atol=1e-3, rtol=1e-3))
+
+    def test_load_regular_lora(self):
+        # This test checks if a regular lora (think of one trained Flux.1 Dev for example) can be loaded
```
Comment (a-r-r-o-w) [inline tests/lora/test_lora_layers_flux.py:586]:
```suggestion
        # This test checks if a regular lora (think of one trained on Flux.1 Dev for example) can be loaded
```

**Review by a-r-r-o-w — APPROVED** [review]
Thanks for making this work @sayakpaul! Just some tiny nits and a question

File: tests/lora/test_lora_layers_flux.py (hunk 4 above)
Comment (sayakpaul) [reply tests/lora/test_lora_layers_flux.py:481]: Exactly!

File: src/diffusers/loaders/lora_pipeline.py [hunk 7]
```diff
@@ -2309,16 +2312,17 @@ def _maybe_expand_transformer_param_shape_or_error_(
...
         for name, module in transformer.named_modules():
             if isinstance(module, torch.nn.Linear):
                 module_weight = module.weight.data
                 module_bias = module.bias.data if module.bias is not None else None
                 bias = module_bias is not None
 
-                lora_A_weight_name = f"{name}.lora_A.weight"
-                lora_B_weight_name = f"{name}.lora_B.weight"
-                if lora_A_weight_name not in state_dict.keys():
+                lora_base_name = name.replace(".base_layer", "") if is_peft_loaded else name
```
Comment (a-r-r-o-w) [inline src/diffusers/loaders/lora_pipeline.py:2322]: Awesome!

File: tests/lora/test_lora_layers_flux.py [hunk 8]
```diff
@@ -546,6 +519,98 @@ def test_lora_expanding_shape_with_normal_lora_raises_error(self):
...
+        lora_output = pipe(**inputs, generator=torch.manual_seed(0))[0]
+
+        normal_lora_A = torch.nn.Linear(in_features, rank, bias=False)
+        normal_lora_B = torch.nn.Linear(rank, out_features, bias=False)
+        lora_state_dict = {
+            "transformer.x_embedder.lora_A.weight": normal_lora_A.weight,
+            "transformer.x_embedder.lora_B.weight": normal_lora_B.weight,
+        }
+
+        pipe.load_lora_weights(lora_state_dict, "adapter-2")
```
Comment (a-r-r-o-w) [inline tests/lora/test_lora_layers_flux.py:564]: Very cool that this would now be possible! :fire:

**Review by a-r-r-o-w — APPROVED** [review]
Thanks Sayak! LGTM

The case for the following can be looked into in a separate PR because it is not too important.
- Step 1: Load N "normal" loras
- Step 2: Load shape expander lora
- Step 3; Run inference

This should currently raise an input size mismatch error on the normal loras when the inputs pass through the normal lora computation path - which is okay. This is because we don't handle the case of rejigging and expanding all the previous N loras if a shape expander lora is loaded after them.

Maybe if someone reports an issue, it would be worth looking into. But otherwise, the recommendation would be to load shape expander lora first, and then all the others loras. Thanks again for making this possible!

**Comment by sayakpaul** [comment]
Just completed running ALL the integration tests for Flux LoRA (`test_lora_layers_flux.py`). All green 🟢
